network/src/dms.rs

```diff
@@ -67,6 +83,133 @@ trait DistributedMessageSetRpcInterface: Send + Sync + 'static {
+            ));
+        }
+        let knowns: HashSet<_> = knowns.into_iter().collect();
+        let messages: Vec<_> = messages
+            .drain(..)
+            .filter(|m| !knowns.contains(&m.to_hash256()))
+            .map(RawMessage::from_message)
+            .collect();
+        Ok(messages)
+    }
+}
+
+async fn read_messages(storage: &impl Storage) -> Result<Vec<Message>, Error> {
+    let files = storage.list_files().await?;
+    let tasks = files
+        .into_iter()
+        .map(|f| async move { storage.read_file(&f).await });
+    let data = future::join_all(tasks)
+        .await
+        .into_iter()
+        .collect::<Result<Vec<_>, _>>()?;
+    let messages = data
+        .into_iter()
+        .map(|d| serde_json::from_str(&d))
+        .collect::<Result<Vec<RawMessage>, _>>()?;
+    let messages = messages
+        .into_iter()
+        .map(|d| d.into_message())
+        .collect::<Result<Vec<_>, _>>()?;
+    Ok(messages)
+}
+
+async fn add_message_but_not_broadcast(
+    storage: &mut impl Storage,
+    message: Message,
+) -> Result<(), Error> {
+    storage
+        .add_or_overwrite_file(
+            &format!("{}.json", message.to_hash256()),
+            message.data().to_owned(),
+        )
+        .await?;
+    Ok(())
+}
+
+async fn read_state(storage: &impl Storage) -> Result<State, Error> {
+    let state: State = serde_json::from_str(&storage.read_file(STATE_FILE_PATH).await?)?;
+    Ok(state)
+}
+
+async fn write_state(storage: &mut impl Storage, state: State) -> Result<(), Error> {
+    storage
+        .add_or_overwrite_file(STATE_FILE_PATH, serde_json::to_string(&state)?)
+        .await?;
+    Ok(())
+}
+
+async fn fetch<S: Storage>(
+    storage: Arc<RwLock<S>>,
+    _network_config: &NetworkConfig,
+    known_peers: &[Peer],
+) -> Result<(), Error> {
+    let mut tasks = Vec::new();
+    let messages = read_messages(&*storage.read().await).await?;
+    let known_messages = messages
+        .into_iter()
+        .map(|m| m.to_hash256())
+        .collect::<Vec<_>>();
+    let state = read_state(&*storage.read().await).await?;
+    let height = state.height;
+
+    for peer in known_peers {
+        let storage = Arc::clone(&storage);
+        let port_key = state.key.clone();
+        let known_messages_ = known_messages.clone();
+        let task = async move {
+            let stub = DistributedMessageSetRpcInterfaceStub::new(Box::new(HttpClient::new(
+                format!(
+                    "http://{}/{}",
+                    peer.address.ip(),
+                    peer.ports
+                        .get(&port_key)
+                        .ok_or_else(|| anyhow!("can't find port key: {}", port_key))?
+                ),
+                reqwest::Client::new(),
+            )));
+            let messages = stub
+                .get_message(height, known_messages_)
+                .await
+                .map_err(|e| anyhow!(e))?
+                .map_err(|e| anyhow!(e))?;
```

Comment:
```suggestion
                .await?
                .map_err(|e| anyhow!(e))?;
```
Let's use a shorter way.

Reply:
updated